pkg/tempdir/tempdir.go

```diff
@@ -0,0 +1,306 @@
+		path := filepath.Join(td.rootDir, entry.Name())
+		if strings.HasPrefix(entry.Name(), TempDirPrefix) {
+			dirs[path] = struct{}{}
+			continue
+		}
+		if strings.HasPrefix(entry.Name(), TempdirLockPrefix) {
+			locks[path] = struct{}{}
+		}
+	}
+
+	out := []staleDir{}
+
+	for path := range dirs {
+		base := filepath.Base(path)
+		lockPath := filepath.Join(td.rootDir, TempdirLockPrefix+base)
+		if _, ok := locks[lockPath]; ok {
+			delete(locks, lockPath)
+		} else {
+			lockPath = ""
+		}
+		out = append(out, staleDir{
+			tempDirPath: path,
+			lockPath:    lockPath,
+		})
+	}
+
+	for path := range locks {
+		out = append(out, staleDir{
+			tempDirPath: "",
+			lockPath:    path,
+		})
+	}
+
+	return out, err
+}
+
+// RecoverStaleDirs identifies and removes stale temporary directories in the root directory.
+// A directory is considered stale if its lock file can be acquired (indicating no active use).
+// The function attempts to remove both the directory and its lock file.
+// Any errors encountered during the recovery process are collected and returned as a single error.
+func (td *TempDir) RecoverStaleDirs() error {
+	potentialStaleDirs, err := td.listPotentialStaleDirs()
+	if err != nil {
+		return fmt.Errorf("error listing potential stale temp dirs in %s: %w", td.rootDir, err)
+	}
+
+	if len(potentialStaleDirs) == 0 {
+		return nil
+	}
+
+	var recoveryErrors []error
+
+	for _, staleDir := range potentialStaleDirs {
+		if staleDir.lockPath != "" {
+			instanceLock, err := lockfile.GetLockFile(staleDir.lockPath)
+			recoveryErrors = append(recoveryErrors, fmt.Errorf("error getting lock file %s: %w", staleDir.lockPath, err))
+			if err := instanceLock.TryLock(); err != nil {
+				logrus.Debugf("Skipping cleanup for locked directory corresponding to %s", staleDir.lockPath)
+				continue
+			}
+		}
+		if staleDir.tempDirPath != "" {
+			if rmErr := os.RemoveAll(staleDir.tempDirPath); rmErr != nil && !os.IsNotExist(rmErr) {
+				recoveryErrors = append(recoveryErrors, fmt.Errorf("error removing stale temp dir %s: %w", staleDir.tempDirPath, rmErr))
+			}
+		}
+
+		if staleDir.lockPath != "" {
+			if err := os.Remove(staleDir.lockPath); err != nil && !os.IsNotExist(err) {
+				recoveryErrors = append(recoveryErrors, fmt.Errorf("error removing stale instance lock file %s: %w", staleDir.lockPath, err))
+			}
+		}
+	}
+
+	return errors.Join(recoveryErrors...)
+}
+
+// initInstanceTempDir creates and locks a new unique temporary directory for this TempDir instance.
+// It acquires the tempDirGlobalLock during its operation to ensure atomicity
+// with other temp dir management operations.
+// The caller should ensure .Cleanup() is called eventually.
+func (td *TempDir) initInstanceTempDir() error {
+	if td.tempDirPath != "" && td.tempDirLock != nil {
+		logrus.Debugf("Temp dir %s already initialized", td.tempDirPath)
+		return nil
+	}
+	randomSuffix := stringid.TruncateID(stringid.GenerateRandomID())
+	tempDirName := TempDirPrefix + randomSuffix
```

Comment:
We would still need to worry about collisions, and probably retry…

I think using `os.CreateTemp` to decide on the lock file path might work well. But, also, see the “leaking memory for staging locks” discussion.